File: solutions/ruby/zebra-puzzle/1/zebra_puzzle.rb
module ZebraPuzzle
    @nationalities = ["EnglishMan", "Spaniard", "Norwegian", "Japanese", "Ukranian"]
    @colors = ["Red", "Green", "Ivory", "Blue", "Yellow"]
    @pets = ["Dog", "Fox", "Snails", "Horse", "Zebra"]
    @drinks = ["Coffee", "Tea", "Milk", "Water", "Orange Juice"]
    @smokes = ["Old Gold", "Kool", "Chesterfield", "Lucky Strike", "Parliament"]

    @results = {}

    def self.water_drinker
        # The Norwegian drinks water.
        if @results[:water_drinker].nil? then
            cache_solve
        end
        @results[:water_drinker]
    end

    def self.zebra_owner
        # The Japanese owns the zebra.
        if @results[:zebra_owner].nil? then
            cache_solve
        end
        @results[:zebra_owner]
    end

    def self.cache_solve
        nat, color, pet, drink, smoke = self.solve
        @results[:zebra_owner] = nat[pet.index("Zebra")]
        @results[:water_drinker] = nat[drink.index("Water")]
    end

    def self.solve
        @nationalities.permutation.each do |n|
          # 10. The Norwegian lives in the first house.
          next if n.index("Norwegian") != 0
    
          @colors.permutation.each do |c|
    
            # 2. The Englishman lives in the red house.
            next if n.index("EnglishMan") != c.index("Red")
    
            # 6. The green house is immediately to the right of the ivory house.
            next if c.index("Ivory") != c.index("Green") + 1
    
            # 15. The Norwegian lives next to the blue house.
            next unless [-1, 1].include?(c.index("Blue") - n.index("Norwegian"))
    
            @pets.permutation.each do |p|
              
              # 3. The Spaniard owns the dog.
              next if n.index("Spaniard") != p.index("Dog")
    
              @drinks.permutation.each do |d|
    
                # 4. Coffee is drunk in the green house.
                next if d.index("Coffee") != c.index("Green")
    
                # 5. The Ukrainian drinks tea.
                next if d.index("Tea") != n.index("Ukranian")
    
                # 9. Milk is drunk in the middle house.
                next if d.index("Milk") != 2
    
                @smokes.permutation.each do |s|
    
                  # 7. The Old Gold smoker owns snails.
                  next if s.index("Old Gold") != p.index("Snails")
    
                  # 8. Kools are smoked in the yellow house.
                  next if s.index("Kool") != c.index("Yellow")
    
                  # 11. The man who smokes Chesterfields lives in the
                  # house next to the man with the fox.
                  next unless [-1, 1].include?(s.index("Chesterfield") - p.index("Fox"))
    
                  # 12. Kools are smoked in the house next to the house
                  # where the horse is kept.
                  next unless [-1, 1].include?(s.index("Kool") - p.index("Horse"))
                  
                  # 13. The Lucky Strike smoker drinks orange juice.
                  next if s.index("Lucky Strike") != d.index("Orange Juice")
    
                  # 14. The Japanese smokes Parliaments.
                  next if s.index("Parliament") != n.index("Japanese")
    
                  return [n, c, p, d, s]
                end
              end
            end
          end
        end
      end
end

File: solutions/javascript/zebra-puzzle/1/zebra-puzzle.js
const colors = ['red', 'green', 'ivory', 'yellow', 'blue'];
const drinks = ['coffee', 'tea', 'milk', 'orange juice', 'water'];
const pets = ['dog', 'snails', 'fox', 'horse', 'zebra'];
const nationalities = ['englishman', 'spaniard', 'ukranian', 'norwegian', 'japanese'];
const smokes = ['old gold', 'kools', 'chesterfields', 'lucky strike', 'parliaments'];

const permutator = (inputArr) => {
  let result = [];

  const permute = (arr, m = []) => {
    if (arr.length === 0) {
      result.push(m)
    } else {
      for (let i = 0; i < arr.length; i++) {
        let curr = arr.slice();
        let next = curr.splice(i, 1);
        permute(curr.slice(), m.concat(next))
     }
   }
 }

 permute(inputArr)

 return result;
}

const PERMUTATIONS = permutator([1,2,3,4,5]);
const rightOf = (a, b) => a === b+1;
const nextTo = (a, b) => rightOf(a, b) || rightOf(b, a);

export class ZebraPuzzle {
  constructor() {
    this.drinkerOfWater = null;
    this.ownerOfZebra = null;
  }

  waterDrinker() {
    if (!this.drinkerOfWater) this.solve();
    return this.drinkerOfWater;
  }

  zebraOwner() {
    if (!this.ownerOfZebra) this.solve();
    return this.ownerOfZebra;
  }


  solve() {
    for (const p of PERMUTATIONS) {
      if (this.solveForColour(p)) break;
    }
  }

  solveForColour(permutation) {
    colors.forEach((color, i) => this[color] = permutation[i]);
    if (
      rightOf(this.green, this.ivory)             // clue 6
    ) {
      for (const p of PERMUTATIONS) {
        if (this.solveForNationality(p))
          return true;
      }
    }
  }

  solveForNationality(permutation) {
    nationalities.forEach((nation, i) => this[nation] = permutation[i]);
    if (
        this.englishman === this.red                 // clue 2
        && this.norwegian === 1               // clue 10
        && nextTo(this.norwegian, this.blue)      // clue 15
    ) {
      this.nationalities = []
      this.nationalities[this.englishman] = "EnglishMan";
      this.nationalities[this.spaniard] = "Spaniard";
      this.nationalities[this.ukranian] = "Ukranian";
      this.nationalities[this.norwegian] = "Norwegian";
      this.nationalities[this.japanese] = "Japanese";
      for (const p of PERMUTATIONS) {
        if (this.solveForBeverages(p))
          return true;
      }
    }
  }

  solveForBeverages(permutation) {
    drinks.forEach((d, i) => this[d] = permutation[i]);
    if (
        this.coffee === this.green                // clue 4
        && this.ukranian === this.tea             // clue 5
        && this.milk === 3                        // clue 9
    ) {
      for (const p of PERMUTATIONS) {
        if (this.solveForSmokes(p))
          return true;
      }
    }
  }


  solveForSmokes(permutation) {
    smokes.forEach((smoke, i) => this[smoke] = permutation[i]);
    if (
        this.kools === this.yellow                // clue 8
        && this['lucky strike'] === this['orange juice']  // clue 13
        && this.japanese === this.parliaments     // clue 14
    ) {
      for (const p of PERMUTATIONS) {
        if (this.solveForPets(p))
          return true;
      }
    }
  }

  solveForPets(permutation) {
    pets.forEach((pet, i) => this[pet] = permutation[i]);
    if (
        this.spaniard === this.dog                 // clue 3
        && this['old gold'] === this.snails           // clue 7
        && nextTo(this.chesterfields, this.fox)   // clue 11
        && nextTo(this.kools, this.horse)         // clue 12
    ) {
      this.drinkerOfWater = this.nationalities[this.water];
      this.ownerOfZebra = this.nationalities[this.zebra];
      return true;
    }
    return false;
  }
}
